### DISCUSSION TITLE
stylix: use with home.file

### DISCUSSION THREAD

**Discussion by henrispriet** [nix-community/stylix#1127] [Ideas]
# Minimal Example

When using stylix with (flake standalone) home-manager I would like to do the following (example for lazygit):
```nix
# home.nix
{
  # ...
  programs.lazygit.enable = true;
  home.file.".config/lazygit/config.yml".source = ./dotfiles/.config/lazygit/config.yml;
  
  stylix = {
    enable = true;
    autoEnable = true;

    polarity = "dark";
    base16Scheme = "${pkgs.base16-schemes}/share/themes/gruvbox-dark-medium.yaml";
  }
  # ...
}
```
```yml
# dotfiles/.config/lazygit/config.yml
gui:
  border: single
notARepository: quit
```

# HomeManager Error

But, I get this error from home-manager:
```
error:
       Failed assertions:
       - Conflicting managed target files: .config/lazygit/config.yml

       This may happen, for example, if you have a configuration similar to

           home.file = {
             conflict1 = { source = ./foo.nix; target = "baz"; };
             conflict2 = { source = ./bar.nix; target = "baz"; };
           }
```

# Attempted Workarounds

## 1
```diff
-  programs.lazygit.enable = true;
```
I've tried leaving out the enable. Then, `home-manager switch` works, but the stylix config isn't applied, only that from the dotfiles (makes sense).

## 2
```diff
-  home.file.".config/lazygit/config.yml".source = ./dotfiles/.config/lazygit/config.yml;
+  home.file.".config/lazygit/config.yml".text = builtins.readFile ./dotfiles/.config/lazygit/config.yml;
```
I've also tried changing the `home.file` from `source` to `text`using `builtins.readFile`, since [nixos wiki mentions](https://nixos.wiki/wiki/Home_Manager#Managing_your_dotfiles) that
> [...] "text" would add the text to the dot file that may also be affected in other places.

But, it gives the same error of `Conflicting managed target files: .config/lazygit/config.yml`.

# Questions

- Is there any way around this?
- Is this a problem with stylix or with home-manager?
- How can this be solved?

**Comment by trueNAHO** [comment]
> - Is this a problem with stylix or with home-manager?

It is an already acknowledged integration limitation between Stylix and Home Manager:

> > I wanted to override a few colors for Gnome (without recreating all the needed derivations), but it looks like that the template is [hardcoded](https://github.com/danth/stylix/blob/6863412636c8f2cb3b7360f747fbd020fbfddf68/modules/gnome/colors.mustache).
>
> Ideally, we would upstream (NixOS, Home Manager, or nix-darwin) relevant options to seamlessly integrate Stylix into external Nix code.
>
> > There's this relevant thread: #159 about moving these templates out-of-tree. Would a PR for separating the mustache template to a different repo be acceptable here?
>
> The [future Vim highlight groups](https://github.com/danth/stylix/issues/534) would allow us to get rid of mustache templates:
>
> > the [roadmap](https://github.com/danth/stylix/issues/534) eventually causes Stylix to implement all templates without relying on external dependencies
> >
> > -- https://github.com/danth/stylix/issues/567#issuecomment-2403246345
>
> [...]
>
> > Are the patches really suitable for upstreaming though? It contains a lot of stylix-specific changes, like removing dark mode. The style sheet itself is also fairly opinionated and custom, which is not something I usually see upstream.
>
> My idea was to move our string injection architecture
>
> ```nix
> xdg.configFile."<MODULE>".text = "<KEY>=<VALUE>";
> ```
>
> upstream, and then we only use its interface:
>
> ```nix
> programs.<MODLE>.extraConfig."<KEY>" = "<VALUE>";
> ```
>
> This centralizes the string injection upstream and hopefully gets rid of override issues:
>
> > the use of `extraConfig` might cause other potentially related issues. As discussed in [#159 (comment)](https://github.com/danth/stylix/issues/159#issuecomment-1937879216), it makes it harder to override options, and as discussed in [#388 (comment)](https://github.com/danth/stylix/pull/388#issuecomment-2126947630), it makes it easier to accidentally override Stylix options.
> >
> > -- https://github.com/danth/stylix/issues/395
>
> -- [NAHO, "Overriding Gnome template"](https://github.com/danth/stylix/issues/621)

> - Is there any way around this?
> - How can this be solved?

Unless you copy-paste and slightly modify the current Stylix module implementation, I am unaware of a reliable solution before the previously mentioned limitations have been resolved. Depending on the configuration format, `lib.mkAfter` could be used.

**Comment by danth** [comment]
Stylix uses the [`programs.lazygit.settings`](https://nix-community.github.io/home-manager/options.xhtml#opt-programs.lazygit.settings) option:

https://github.com/danth/stylix/blob/1d7b70ed9ee4c3b24ed6b0c7c64a0ee5fcc4ae10/modules/lazygit/hm.nix#L16

If you adapt your own configuration to also use this option, the two should merge together seamlessly.

Of course, this means you no longer have a configuration file in the original format to use outside of Home Manager.

**Comment by 0xda157** [comment]
this feels like it should be a discussion not an issue

**Comment by trueNAHO** [comment]
> this feels like it should be a discussion not an issue

This issue covers a real tangible Stylix limitation that is planned to be resolved in the future. Since it is also listed in the Roadmap, it seems better to not convert this issue to a discussion and consider this issue somewhat as a tracking issue.

**Comment by Flameopathic** [comment]
this seems very related to #481. 

> This issue covers a real tangible Stylix limitation that is planned to be resolved in the future.

is our plan of resolving this the same as in #481 as well, which is upstreaming options?

**Comment by 0xda157** [comment]
> > This issue covers a real tangible Stylix limitation that is planned to be resolved in the future.
> 
> is our plan of resolving this the same as in [#481](https://github.com/danth/stylix/issues/481) as well, which is upstreaming options?

This is a home-manager issue, the only solutions (I can think of) are either not using `programs.<name>` at all or upstreaming extraConfig options. 

> > this feels like it should be a discussion not an issue
> 
> This issue covers a real tangible Stylix limitation that is planned to be resolved in the future. Since it is also listed in the Roadmap, it seems better to not convert this issue to a discussion and consider this issue somewhat as a tracking issue.

The reason I think this should be a discussion is because there isn't really a defined use case aside from "I want to use `home.file`". Generally you can just read from json and dump that into the settings option (in the case of lazygit you would have to convert your config to json first). The only way I can see to resolve this downstream would be re-implementing the functionality of home-manager's `programs.*` for each target which isn't viable.

**Comment by trueNAHO** [comment]
> > This issue covers a real tangible Stylix limitation that is planned to be resolved in the future.
>
>
> is our plan of resolving this the same as in [#481](https://github.com/danth/stylix/issues/481) as well, which is upstreaming options?

AFAIK, yes.

> > > this feels like it should be a discussion not an issue
> >
> >
> > This issue covers a real tangible Stylix limitation that is planned to be resolved in the future. Since it is also listed in the Roadmap, it seems better to not convert this issue to a discussion and consider this issue somewhat as a tracking issue.
>
> The reason I think this should be a discussion is because there isn't really a defined use case aside from "I want to use `home.file`". Generally you can just read from json and dump that into the settings option (in the case of lazygit you would have to convert your config to json first). The only way I can see to resolve this downstream would be re-implementing the functionality of home-manager's `programs.*` for each target which isn't viable.

I am fine with converting this issue into a discussion since we already have https://github.com/danth/stylix/issues/481.
